instrumentation/servlet/servlet-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/ServletHelper.java

```diff
@@ -82,54 +77,47 @@ public boolean mustEndOnHandlerMethodExit(REQUEST request) {
   }
 
   /**
-   * Response object must be attached to a request prior to {@link
-   * #attachAsyncListener(ServletRequestContext)} being called, as otherwise in some environments it
-   * is not possible to access response from async event in listeners.
+   * Response object must be attached to a request prior to {@link #attachAsyncListener(REQUEST,
+   * Context)} being called, as otherwise in some environments it is not possible to access response
+   * from async event in listeners.
    */
-  public void setAsyncListenerResponse(REQUEST request, RESPONSE response) {
-    accessor.setRequestAttribute(request, ASYNC_LISTENER_RESPONSE_ATTRIBUTE, response);
+  public void setAsyncListenerResponse(Context context, RESPONSE response) {
+    ServletAsyncContext.setAsyncListenerResponse(context, response);
   }
 
-  public RESPONSE getAsyncListenerResponse(REQUEST request) {
-    @SuppressWarnings("unchecked")
-    RESPONSE response =
-        (RESPONSE) accessor.getRequestAttribute(request, ASYNC_LISTENER_RESPONSE_ATTRIBUTE);
-    return response;
+  @SuppressWarnings("unchecked")
+  public RESPONSE getAsyncListenerResponse(Context context) {
+    return (RESPONSE) ServletAsyncContext.getAsyncListenerResponse(context);
   }
 
-  public void attachAsyncListener(REQUEST request) {
-    ServletRequestContext<REQUEST> requestContext = new ServletRequestContext<>(request, null);
-    attachAsyncListener(requestContext);
-  }
-
-  private void attachAsyncListener(ServletRequestContext<REQUEST> requestContext) {
-    REQUEST request = requestContext.request();
-    Context context = getServerContext(request);
```

Comment:
Existing code was wrong here, we don't add context as request attribute on all servers so `getServerContext` returns null there.